File: BookServiceJava/src/test/java/org/elcata98/mybooks/booksservice/mvc/BookControllerMvcTest.java
package org.elcata98.mybooks.booksservice.mvc;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.elcata98.mybooks.booksservice.model.Book;
import org.elcata98.mybooks.booksservice.response.Response;
import org.elcata98.mybooks.booksservice.validator.EntityValidator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import java.time.LocalDateTime;
import java.time.ZoneOffset;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
@SpringBootTest
@ContextConfiguration
public class BookControllerMvcTest {

    @Autowired
    private WebApplicationContext context;

    private MockMvc mockMvc;

    private ObjectMapper objectMapper = new ObjectMapper();

    private Book book;

    private final TypeReference<Response<Book>> responseTypeReference = new TypeReference<>() {};

    @BeforeEach
    void setUp() {

        long now = LocalDateTime.now().toEpochSecond(ZoneOffset.UTC);

        mockMvc = MockMvcBuilders.webAppContextSetup(context).build();

        book = new Book();
        book.setTitle("Title - " + now);
        book.setAuthor("Author - " + now);
        book.setLanguage("Catalan - " + now);
    }

    @Test
    void testCreate() throws Exception {

        createBook();
    }

    @Test
    void testCreateNoMandatoryFields() throws Exception {

        book = new Book();

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .post("/books")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isBadRequest());
    }

    @Test
    void testCreateWithId() throws Exception {

        book.generateId();

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .post("/books")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isBadRequest())
                .andDo(mvcResponse ->
                        assertTrue(mvcResponse.getResponse().getContentAsString().contains(EntityValidator.CREATE_ERROR_MSG))
                );
    }

    @Test
    void testGet() throws Exception {

        book = createBook();

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .get("/books/" + book.getId())
                                .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andDo(mvcResult -> assertFalse(mvcResult.getResponse().getContentAsString().isEmpty()))
                .andDo(mvcResult ->
                        assertEquals(
                                book,
                                objectMapper.readValue(mvcResult.getResponse().getContentAsString(), responseTypeReference).getResponse()
                        )
                );
    }

    @Test
    void testGetNotFound() throws Exception {

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .get("/books/not_exists")
                                .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isNotFound());
    }

    @Test
    void testUpdate() throws Exception {

        book = createBook();
        book.setNumVolumes((byte) 10);

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .put("/books/" + book.getId())
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk())
                .andDo(mvcResult -> assertFalse(mvcResult.getResponse().getContentAsString().isEmpty()))
                .andDo(mvcResult ->
                        assertEquals(
                                book,
                                objectMapper.readValue(mvcResult.getResponse().getContentAsString(), responseTypeReference).getResponse()
                        )
                );
    }

    @Test
    void testUpdateMismatch() throws Exception {

        book = createBook();
        book.setNumVolumes((byte) 10);

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .put("/books/mismatch")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isBadRequest());
    }

    @Test
    void testDelete() throws Exception {

        book = createBook();

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .delete("/books/" + book.getId()))
                .andExpect(status().isNoContent());
    }

    @Test
    void testDeleteError() throws Exception {

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .delete("/books/error"))
                .andExpect(status().is5xxServerError());
    }

    private Book createBook() throws Exception {

        Book[] books = new Book[1];

        mockMvc
                .perform(
                        MockMvcRequestBuilders
                                .post("/books")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isCreated())
                .andExpect(header().exists("Location"))
                .andDo(mvcResult -> assertFalse(mvcResult.getResponse().getContentAsString().isEmpty()))
                .andDo(mvcResult ->
                        books[0] =
                                objectMapper.readValue(mvcResult.getResponse().getContentAsString(), responseTypeReference).getResponse()
                );

        return books[0];
    }
}

File: BooksService/src/test/kotlin/org/elcata98/mybooks/booksservice/mvc/BookControllerMvcTest.kt
package org.elcata98.mybooks.booksservice.mvc

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.elcata98.mybooks.booksservice.model.Book
import org.elcata98.mybooks.booksservice.response.Response
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultMatcher
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.header
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext
import java.time.LocalDateTime
import java.time.ZoneOffset


@SpringBootTest
@ContextConfiguration
class BookControllerMvcTest {

    @Autowired
    private val context: WebApplicationContext? = null

    private var mockMvc: MockMvc? = null

    private val objectMapper: ObjectMapper = ObjectMapper()

    @BeforeEach
    fun setUp() {
        mockMvc = MockMvcBuilders.webAppContextSetup(context!!).build();
    }

    @Test
    fun testCreate() {

        createBook()
    }

    @Test
    fun testCreateWithId() {

        val book = Book()
        val now = LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)
        book.title = "Title - $now"
        book.author = "Author - $now"
        book.language = "Catalan - $now"
        book.generateId()

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .post("/books")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isBadRequest)
    }

    @Test
    fun testGet() {

        val book = createBook()

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .get("/books/" + book.bookId)
                                .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isOk)
                .andExpect(ResultMatcher { mvcResult -> mvcResult.response.contentAsString.isNotEmpty() })
                .andDo { mvcResult ->
                    assertEquals(book, objectMapper.readValue(mvcResult.response.contentAsString, object : TypeReference<Response<Book>>() {}).response!!)
                }
    }

    @Test
    fun testGetNotFound() {

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .get("/books/not_exists")
                                .contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().isNotFound)
    }

    @Test
    fun testUpdate() {

        val book = createBook()
        book.numVolumes = 10

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .put("/books/" + book.bookId)
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andDo { mvcResult -> println(mvcResult.response.contentAsString) }
                .andExpect(status().isOk)
                .andExpect(ResultMatcher { mvcResult -> mvcResult.response.contentAsString.isEmpty() })
                .andDo { mvcResult ->
                    assertEquals(book, objectMapper.readValue(mvcResult.response.contentAsString, object : TypeReference<Response<Book>>() {}).response!!)
                }
    }

    @Test
    fun testUpdateIdMismatch() {

        val book = createBook()
        book.numVolumes = 10

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .put("/books/mismatch")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andDo { mvcResult -> println(mvcResult.response.contentAsString) }
                .andExpect(status().isBadRequest)
    }

    @Test
    fun testDelete() {

        val book = createBook()

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .delete("/books/" + book.bookId)
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isNoContent)
                .andExpect(ResultMatcher { mvcResult -> mvcResult.response.contentAsString.isEmpty() })
    }

    @Test
    fun testDeleteError() {

        val book = createBook()

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .delete("/books/error")
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().is5xxServerError)
    }

    private fun createBook(): Book {

        var book = Book()
        val now = LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)
        book.title = "Title - $now"
        book.author = "Author - $now"
        book.language = "Catalan - $now"

        mockMvc!!
                .perform(
                        MockMvcRequestBuilders
                                .post("/books")
                                .content(objectMapper.writeValueAsString(book))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON))
                .andExpect(status().isCreated)
                .andExpect(header().exists("Location"))
                .andExpect(ResultMatcher { mvcResult -> mvcResult.response.contentAsString.isNotEmpty() })
                .andDo { mvcResult ->
                    book = objectMapper.readValue(mvcResult.response.contentAsString, object : TypeReference<Response<Book>>() {}).response!!
                }

        return book;
    }
}